### DISCUSSION TITLE
Multiple updates via a single query

### DISCUSSION THREAD

**Discussion by Fuuzetsu** [diesel-rs/diesel#2879] [Ideas]
This continues on from https://github.com/diesel-rs/diesel/issues/1517

Basically, it is currently impossible(?) to do arbitrary updates to multiple distinct rows with a single query. As an imaginary example to demonstrate what I mean, let's suppose we're selling tickets to some venue that allocates seats to the customers at the end of the ticket period rather than up-front. So:

* many customers reserve tickets to the venue
* time passes or tickets run out
* a piece of software allocates seats to every customer

Let's naively model it with a single table for illustration purposes

```sql
CREATE TABLE customers ( customer_id bigint PRIMARY KEY…, seat_id bigint, <more stuff here, doesn't matter> );
```

When customer buys a ticket, they get put on in `customers` with their personal data but `seat_id` stays NULL.

After time passes, we now have some software allocate a seat_id to every customer.

Currently with diesel, we can use `AsChangeset` to update a single customer, maybe it'd look something like

```rust
#[derive(AsChangeset)]
#[table_name = "customers"]
pub struct AllocateSeat {
  pub customer_id: int64,
  pub seat_id: int64
}
```

we should be able to now update a single customer like so:
```rust
let allocation: AllocateSeat = <magical piece of code to decide seat>;
diesel::update(customers::table).set(&allocation).execute(conn)
```

So far so good, but what happens if our venue has 50_000 seats? The code below doesn't compile:

```rust
let allocations: Vec<AllocateSeat> = <magical code to decide all seats>;
diesel::update(customers::table).set(&allocations).execute(conn)
```

So we're force to instead loop over `allocations` and run `update` for each element. This means 50_000 small queries.

At the very least postgres is able to do this in one (albeit large) query: see [this StackOverflow](https://stackoverflow.com/questions/18797608/update-multiple-rows-in-same-query-using-postgresql) for an example. I believe that crux is the `FROM (VALUES …)` where you can put all the values you want at once and then execute them with `UPDATE` and `SET` as usual.

To answer the questions [from the original ticket](https://github.com/diesel-rs/diesel/issues/1517):

>What should exactly be added to diesel?

Ability to update all the rows specified in one query if the underlying DB supports it.

>How would the corresponding public API look?

I'm not too in-depth to all the APIs but my naive suggestion would be to have a trait like `impl<T: AsChangeset> AsChangeset for &[T]` or something along these lines. Maybe we have to make use of `.values`, unsure.

>Which backends support this SQL variant.

I only know of postgres so I can't comment on others. Quick search shows [sqlite should be able to do it](https://stackoverflow.com/questions/11563869/update-multiple-rows-with-different-values-in-a-single-sql-query) as well as [mysql](https://stackoverflow.com/a/34866431).

Where it's not easily supported, diesel could just fall back to doing a loop with one query per update. That would give consistent API at least.

>Is this supported for all versions of a backend or only for specific versions

I don't know the exact versions but I think this has been a thing in mysql and postgres for many years now, definitely not a new feature.

>Do other backends support the same feature with different syntax?

See above for some links.

> *Are there any invariants, which should be checked at compile time?

I don't think there's anything that's not already done with the vanilla `AsChangeset`

**Comment by weiznich** [comment]
According to your description this can be implemented without any breaking change, which means it's not something I will work on at least till 2.0 is released (That release already took too long, so I don't want to block it on any more features).
I would likely accept a PR implementing this as long as it contains the following things:
* The corresponding change to allow updating multiple values at once. This should not break any existing API
* Updated documentation showing this feature in our API documentation
* Some test cases showing that this feature works
* Maybe a compile fail test showing that this feature is only implemented for a specific backend?

**Comment by FayCarsons** [comment]
Has this been implemented? Currently dealing w/ the same problem and am not seeing a solution in the docs, though I'm a bit of a SQL noob so I could be missing something.

**Reply by weiznich** [reply]
@FayCarsons This feature is still waiting on someone that picks it up and implements it. Your contributions are welcome here.

**Comment by kristof-mattei** [comment]
I've been noodling on this issue for a while and I've been studying the source code to see if I could come up with a solution.

As stated above, idea would be to pass on a `&[T]` to `.set`.


As such we need to implement `AsChangeset`: 

```
impl<'t, T> AsChangeset for &'t [T]
where
    &'t T: AsChangeset,
{
    type Target = <&'t T as AsChangeset>::Target;

    // Option because the list might be empty, and then we don't generate a changeset
    type Changeset = Option<AssignMany<T>>;

    fn as_changeset(self) -> Self::Changeset {
        self.get(0).map(|t| AssignMany<?, ?> {
            // assignments: t.as_changeset().collect(),
        })
    }
}

#[derive(Debug, Clone, QueryId)]
pub struct AssignMany<Left, Right> {
    // Left & Right might be able to be fixed to 'column' or something like that.
    assignments: Vec<(Left, Right)>,
}
```

Now, the changeset for `&[T]` that we want to appear in the `SET` clause is the one from `T`, but with the `Right` replaced with a temporary column name.

And to do this replacement my current understand is that I need to iterate over the n `eq`s generated by the `AsChangeset` macro, and replace the right-hand side by a temporary column name.

However the macro generates a tuple, and I have not found a way yet to iterate over them.

I'm going to keep looking, but I'd appreciate pointers (or maybe I'm thinking of this all wrong?)!

**Reply by weiznich** [reply]
I think it's the correct way to introduce new wrapping types there. The `AssignMany` type could just wrapp the `&[T]` there first, we can improve on that later on. That transformation is mostly there to desuger structs implementing `AsChangeset` to a tuple. 

Based on that we can likely specialize [this impl here](https://github.com/diesel-rs/diesel/blob/7d8844547498407b6d5b8d1a37e695d35aa8c08b/diesel/src/query_builder/update_statement/mod.rs#L191-L214) to generate different SQL for the batch and not batch variant. From there on it's "just" destructing the types again and again to get the right information out there. After all everything is in there via different constraints. I hope that helps at least a bit to understand where to start looking.

**Reply by be-mevsane** [reply]
Hi @kristof-mattei, did you have any chance to further pursue your attempt? I'd like to help with what little I can contribute, if you don't mind.

**Reply by weiznich** [reply]
@be-mevsane There was no PR or similar for this, so we still look for a contributor that implements this.

**Reply by be-mevsane** [reply]
@weiznich Hi and thanks for the reply!
I'll go ahead start off with a feature branch locally. Will eventually push it to the repo after confirming with you at a later time, if that's ok?
Will probably call the branch something like `feature_multiple_updates_via_single_query`.

**Reply by be-mevsane** [reply]
Hi @weiznich,
I have set up a branch [feature_batch_update](https://github.com/be-mevsane/diesel/tree/feature_batch_update) with initial setup and some documentation.


You can find my [understanding](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L158) of how the Marco to sql statement generation flows and a vaguely prepared [todo](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L223) list.
I would appreciate your opinion on it. Is that the direction you imagined?
 
---

**The following is unclear to me and needs clarification please:**

I understand that in [diesel_for_each_tuple ](https://github.com/diesel-rs/diesel/blob/main/diesel_derives/src/diesel_for_each_tuple.rs#L13) all combinations for the tuples (0...MAX_TUPLE_SIZE) use the macro [tuple_impls](https://github.com/diesel-rs/diesel/blob/main/diesel/src/type_impls/tuples.rs#L37):

Counting by the index (please forgive my off by one error, if I made one):
Tuple with no elements like () ?
`tuple_impls{ 1 { (0) -> T, ST, TT } }`
Tuple with one element like (one) ?
`tuple_impls{ 2 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, } }`
Tuple with two elements like (one, two) ?
`tuple_impls{ 3 { (0) -> T, ST, TT, (1) -> T1, ST1, TT1, (2) -> T2, ST2, TT2, } }`
and so on until MAX_TUPLE_SIZE is reached.

I could not figure out the meaning of $T:ident and $ST:ident in 'tuple_impls' for each tuple element.
How does a tuple translate to T, ST and TT ? (maybe TT is unimportant since unused)
Maybe:
```
impl<$($T,)+ Tab> for (Eq<Table::a, example.a>, Eq<Table::b, example.b>)
T  = type representing the individual elements Eq<Left, Right>

impl<$($T,)+ $($ST,)+ Tab> for (Eq<Table::a, example.a>, Eq<Table::b, example.b>)
T  = type of Left for all Eq<Left, Right>
ST = type of Right for all Eq<Left, Right>
```
Or:
```
T  = always the type for Eq<Table::a, example.a>
ST = stands for Bound<Text, &String> since example.a is of type String
```

I'm just guessing at this point and would appreciate some explanation...

**Reply by weiznich** [reply]
> I could not figure out the meaning of $T:ident and $ST:ident in 'tuple_impls' for each tuple element.

Basically `$T`, `$ST` and `$TT` are just arbitrary names there. In the macro they are identifiers generated by the outer proc macro so that the inner macro can construct tuples based on them, by e.g. using them as generics. 

Usually `$T` refers to the rust side type, `$ST` to the SQL side type (mostly from `diesel::sql_types` and `$TT` is there because sometimes you need another third generic type. What exactly these types refer to is different from impl to impl. Hopefully that helps a bit.

> vaguely prepared [todo](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L223) list.

That sounds reasonable. One thing we also need to take care of is not to break backward compatibility. You step 3 sounds a bit dangerous in that regard, maybe we can just do the "enum" at type by having sequences of values rather be a different type. After all there is nothing that specifies that `AsChangeset::Changeset` is always the same type. That means for `Vec<C: AsChanegset>` let the type just be `BatchChangeSet<C::Changeset>` while for plain changesets nothing changes.

**Reply by be-mevsane** [reply]
> usually `$T` refers to the rust side type, `$ST` to the SQL side type (mostly from `diesel::sql_types` ...

Thanks, that helps indeed! I'll go back to `Insertable` as an example and maybe understand what's happening there properly.

> maybe we can just do the "enum" at type by having sequences of values rather be a different type

Point taken, I can do that. Probably misinterpreted your previous comment:

> Based on that we can likely specialize [this impl here](https://github.com/diesel-rs/diesel/blob/main/diesel/src/query_builder/update_statement/mod.rs#L189-L212) to generate different SQL for the batch and not batch variant.

My intention was not to introduce breaking changes with my proposal. I thought of a way to generate a different SQL at [this impl here](https://github.com/diesel-rs/diesel/blob/main/diesel/src/query_builder/update_statement/mod.rs#L189-L212) since `UpdateStatement::where_clause` becomes susceptible to change when targeting a new SQL like
`UPDATE tab SET .. FROM ( VALUES .. ) AS tmp(col_a, col_b, col_c) WHERE tab.col_b = tmp.col_b;`.

Maybe we can introduce `BatchUpdateStatement` with specialized behaviour that gets returned by calling `UpdateStatement::set_batch` similar to UpdateStatement being returned when calling [`UpdateStatement::set`](https://github.com/diesel-rs/diesel/blob/main/diesel/src/query_builder/update_statement/mod.rs#L35-L47). At first glance forking there kinda ensures backward compatibility. Other problems may arise though.

I'll just think about other options when I get there. =)

**Reply by be-mevsane** [reply]
I now have a first working test !!!
`cargo test -p diesel_derives -F diesel/postgres -F diesel_derives/postgres named_struct_batch -- --nocapture`

> That means for Vec<C: AsChanegset> let the type just be BatchChangeSet<C::Changeset> while for plain changesets nothing changes.

Hope I understood your suggestion right with [my attempt](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L285-L293).

Since you want to keep [`Eq<L,R>::Changeset = Assign`](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L61) unchanged, it becomes a matter of getting the info out of a vector containing tuples of Assign structs: `Vec<(Assign, Assign, ...)>`.

I added additional [`BatchUpdate traits`](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/batch_update.rs#L23-L34) which allow getting only the column name and/or value when going through a `QueryFragment` pass. Original behavior of `Assign` is preserved.

All came together when implementing `QueryFragment` for [`BatchChangeSet`](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel/src/query_builder/update_statement/changeset.rs#L458). All steps building the `batch update query` are documented here too.

Lastly updating the [test function](https://github.com/be-mevsane/diesel/blob/feature_batch_update/diesel_derives/tests/as_changeset.rs#L1006) and it worked.

I would appreciate your opinion again before I continue.

**Reply by weiznich** [reply]
That already looks really good. Congratulations to get this working :tada:

I suggest opening a draft PR with the current state, as that makes it much easier to put comments on particular code parts and also to keep track of what needs to be implemented.

**Reply by be-mevsane** [reply]
[Draft PR](https://github.com/diesel-rs/diesel/pull/4916) created 🎉
